### PR TITLE
switch to RoleBindingRestrictionInformer for restrictusers admission

### PR DESCRIPTION
**Pull request by sallyom**
Bug 1706637
1) use RBRInformer rather than getter for admission, to cache rbrs/improve performance
2) add extended tests for RBRs
 
@enj please review

### CODE REVIEW

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go [hunk 1]
```diff
@@ -167,19 +163,17 @@ func (q *restrictUsersAdmission) Validate(a admission.Attributes) (err error) {
...
 
-	// TODO: Cache rolebinding restrictions.
-	roleBindingRestrictionList, err := q.roleBindingRestrictionsGetter.RoleBindingRestrictions(ns).
-		List(metav1.ListOptions{})
+	roleBindingRestrictionList, err := q.rbrLister.List(labels.Everything())
 	if err != nil {
 		return admission.NewForbidden(a, err)
 	}
-	if len(roleBindingRestrictionList.Items) == 0 {
+	if len(roleBindingRestrictionList) == 0 {
```
Comment (enj) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: need a `HasSynced` somewhere (see where other admission plugins do that)

File: pkg/cmd/server/admission/types.go [hunk 2]
```diff
@@ -59,3 +60,8 @@ type WantsUserInformer interface {
 	SetUserInformer(userinformer.SharedInformerFactory)
 	admission.InitializationValidator
 }
+
+type WantsAuthorizationInformer interface {
+	SetAuthorizationInformer(authorizationinformer.RoleBindingRestrictionInformer)
```
Comment (enj) [inline pkg/cmd/server/admission/types.go]: You want the whole authz interface ?

File: pkg/cmd/server/admission/init.go [hunk 3]
```diff
@@ -23,6 +24,7 @@ type PluginInitializer struct {
 	RegistryHostnameRetriever    registryhostname.RegistryHostnameRetriever
 	SecurityInformers            securityv1informer.SecurityContextConstraintsInformer
 	UserInformers                userinformer.SharedInformerFactory
+	AuthorizationInformers       authorizationinformer.RoleBindingRestrictionInformer
```
Comment (enj) [inline pkg/cmd/server/admission/init.go]: same interface comment

**Comment by sallyom** [comment]
_fixing up test code now_

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go [hunk 4]
```diff
@@ -118,6 +121,10 @@ func subjectsDelta(elementsToIgnore, elements []rbac.Subject) []rbac.Subject {
 // in the namespace.
 func (q *restrictUsersAdmission) Validate(a admission.Attributes) (err error) {
 
+	if !q.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
```
Comment (enj) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: this should be after the checks that skip this admission plugin

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go [hunk 5]
```diff
@@ -43,17 +45,23 @@ type GroupCache interface {
 type restrictUsersAdmission struct {
 	*admission.Handler
 
-	roleBindingRestrictionsGetter authorizationtypedclient.RoleBindingRestrictionsGetter
-	userClient                    userclient.Interface
-	kubeClient                    kubernetes.Interface
-	groupCache                    GroupCache
+	rbrLister  authorizationv1listers.RoleBindingRestrictionLister
+	rbrSynced  func() bool
```
Comment (enj) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: `rbrSynced cache.InformerSynced`

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go [hunk 6]
```diff
@@ -186,6 +189,10 @@ func NewInformers(versionedInformers clientgoinformers.SharedInformerFactory, lo
 	if err != nil {
 		return nil, err
 	}
+	authzClient, err := authzclient.NewForConfig(jsonLoopbackClientConfig)
```
Comment (enj) [inline pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: this does make it so that all future authz informers will end up using JSON.  just something to think about

**Comment by stlaz** [comment]
To fix the unit test, apply
```diff
From 22943028fcc3d8a7e16607dcbf8b55ddf2de2d7a Mon Sep 17 00:00:00 2001
From: Stanislav Laznicka <slaznick@redhat.com>
Date: Tue, 7 May 2019 13:50:29 +0200
Subject: [PATCH] fix unit tests

---
 .../apiserver/admission/restrictusers/restrictusers_test.go   | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go b/pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go
index 1fa418ba4d..008ee15adf 100644
--- a/pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go
+++ b/pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go
@@ -372,7 +372,9 @@ func TestAdmission(t *testing.T) {
 		kclientset := fake.NewSimpleClientset(tc.kubeObjects...)
 		fakeUserClient := fakeuserclient.NewSimpleClientset(tc.userObjects...)
 		indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
-		indexer.Add(tc.rbrs)
+		for _, rbr := range tc.rbrs {
+			indexer.Add(rbr)
+		}
 		rbrLister := authorizationv1listers.NewRoleBindingRestrictionLister(indexer)
 		plugin := newTestAdmission(rbrLister, kclientset, fakeUserClient, fakeGroupCache{})
 
-- 
2.21.0
```

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go [hunk 7]
```diff
@@ -356,19 +371,12 @@ func TestAdmission(t *testing.T) {
...
-		if err != nil {
-			t.Errorf("unexpected error initializing admission plugin: %v", err)
-		}
-
-		plugin.(*restrictUsersAdmission).kubeClient = kclientset
-		plugin.(*restrictUsersAdmission).roleBindingRestrictionsGetter = fakeAuthorizationClient.AuthorizationV1()
-		plugin.(*restrictUsersAdmission).userClient = fakeUserClient
-		plugin.(*restrictUsersAdmission).groupCache = fakeGroupCache{}
+		indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+		indexer.Add(tc.rbrs)
```
Comment (stlaz) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go]:
```go
		for _, rbr := range tc.rbrs {
			indexer.Add(rbr)
		}
```

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go (hunk 5 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: thanks : ) got it

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go (hunk 4 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: right, done

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go (hunk 1 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: added

**Comment by sallyom** [comment]
addressed comments, pushed, but still broken, still working on this.

File: test/integration/restrictusers_test.go [hunk 8]
```diff
@@ -16,20 +16,12 @@ import (
 )
 
 func TestRestrictUsers(t *testing.T) {
-	masterConfig, err := testserver.DefaultMasterOptions()
+	masterConfig, clusterAdminKubeConfig, err := testserver.StartTestMaster()
```
Comment (stlaz) [inline test/integration/restrictusers_test.go]: Since RBRs are now by default registered as an admission plugin, shouldn't `testutil.WaitForRoleBindingRestrictionCRDAvailable()` be called after kube-apiserver has started?

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go (hunk 7 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers_test.go]: doh! : ) thnx

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go [hunk 9]
```diff
@@ -134,6 +135,10 @@ func (q *restrictUsersAdmission) Validate(a admission.Attributes) (err error) {
 		return nil
 	}
 
+	if !q.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
```
Comment (enj) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: this should be right before the `rbrLister.List` call - the rest of the checks do not need RBRs to be functional

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go [hunk 10]
```diff
@@ -167,19 +172,17 @@ func (q *restrictUsersAdmission) Validate(a admission.Attributes) (err error) {
 		return nil
 	}
 
-	// TODO: Cache rolebinding restrictions.
-	roleBindingRestrictionList, err := q.roleBindingRestrictionsGetter.RoleBindingRestrictions(ns).
-		List(metav1.ListOptions{})
+	roleBindingRestrictionList, err := q.rbrLister.List(labels.Everything())
 	if err != nil {
 		return admission.NewForbidden(a, err)
```
Comment (enj) [inline pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: while you are here, change all of the `return admission.NewForbidden(a, err)` to be something like `return admission.NewForbidden(a, fmt.Errorf("<useful message here>: %v", err))`

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go [hunk 11]
```diff
@@ -95,6 +95,8 @@ func NewDefaultOffPluginsFunc(kubeDefaultOffAdmission sets.String) func() sets.S
 		kubeOff.Delete(additionalDefaultOnPlugins.List()...)
 		kubeOff.Delete(openshiftAdmissionPluginsForKube...)
 		kubeOff.Delete(customresourcevalidationregistration.AllCustomResourceValidators...)
+		// TODO: Why is this necessary?
+		kubeOff.Insert("authorization.openshift.io/RestrictSubjectBindings")
```
Comment (sallyom) [inline pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: @stlaz adding this back in (didn't realize it was removed in the RBR->CRD previous PRs).  Why this is necessary I do not know, but when added things work.  Must be a good explanation but I'm at a loss.  (Why is this necessary for RBRs but not SCCs or CRQs?) :woman_shrugging:

Comment (enj) [reply pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: If you add this back, you must also have an e2e test with RBR to confirm that is working at all.

File: test/integration/restrictusers_test.go (hunk 8 above)
Comment (sallyom) [reply test/integration/restrictusers_test.go]: I've added it.

File: test/util/server/server.go [hunk 12]
```diff
@@ -333,6 +333,7 @@ func StartConfiguredMasterWithOptions(masterConfig *configapi.MasterConfig, stop
 
 	// SCCs are required for OpenShift apiserver bootstrap process
 	util.WaitForSecurityContextConstraintsCRDAvailable(clientConfig)
+	util.WaitForRoleBindingRestrictionCRDAvailable(clientConfig)
```
Comment (enj) [inline test/util/server/server.go]: these waits need to error out

Comment (enj) [reply test/util/server/server.go]: (or take `t` as input and `t.Fatal`)

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go (hunk 10 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: will do, I'll let tests run w/ this iteration first. : )

File: test/util/server/server.go (hunk 12 above)
Comment (sallyom) [reply test/util/server/server.go]: right, done

File: pkg/authorization/apiserver/admission/restrictusers/restrictusers.go (hunk 9 above)
Comment (sallyom) [reply pkg/authorization/apiserver/admission/restrictusers/restrictusers.go]: done

**Comment by enj** [comment]
> this does make it so that all future authz informers will end up using JSON. just something to think about

Thinking on this some more - we should assume that we will forget about this JSON thing in the future.  So instead of having this set up "authorization informers," have it set up _only_ RBR informers (with a comment explaining that if you want other authorization informers you need to set up a new one that uses proto).

File: pkg/cmd/server/admission/types.go (hunk 2 above)
Comment (sallyom) [reply pkg/cmd/server/admission/types.go]: switched back to just RBRInformer

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go (hunk 11 above)
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: I did not add it back, but I added e2e anyways : )

**Comment by sallyom** [comment]
terraform errors, retesting

**Comment by sallyom** [comment]
trying to defeat, I'm assuming since that ns is deleted we're in the clear
```
level=error msg="\t* module.dns.aws_route53_record.api_external: 1 error occurred:"
level=error msg="\t* aws_route53_record.api_external: [ERR]: Error building changeset: InvalidChangeBatch: [Tried to create resource record set [name='api.ci-op-m37bjxmf-55c01.origin-ci-int-aws.dev.rhcloud.com.', type='A'] but it already exists]"
level=error msg="\tstatus code: 400, request id: a594104e-73f0-11e9-a8f8-2f881a54a7ab"
```

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go (hunk 6 above)
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: there's a comment above explaining that CRDs require JSON

File: test/extended/authorization/rolebinding_restrictions.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+	rbacv1 "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/apis/authorization"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("[Feature: RoleBinding Restrictions][Serial] RoleBindingRestrictions should be functional", func() {
```
Comment (sallyom) [inline test/extended/authorization/rolebinding_restrictions.go]: Added [Serial] bc when run parallel random test cases fail (not same ones every time).  I need to investigate this more.

Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: With my local cluster of the day, removed 'Serial' and tests are passing.  I'm seeing `err:"net/http: request canceled (Client.Timeout exceeded while reading body)"` however, so maybe w/ parallel tests will be flaky? running through CI to investigate.

Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: passed thru CI as parallel,once this merges i'll monitor for flakes

**Comment by sallyom** [comment]
@enj ready for a merge? thanks

File: test/extended/authorization/rolebinding_restrictions.go [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+				user2 := "eve"
+				_, err := oc.AdminAuthorizationClient().Authorization().Roles(ns).Create(generateRole(ns))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindings(ns).Create(generateRolebinding(ns, user1, "rb1"))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindingRestrictions(ns).Create(generateAllowUserRolebindingRestriction(ns, user1))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindings(ns).Create(generateRolebinding(ns, user2, "rb2"))
+				o.Expect(err).To(o.HaveOccurred())
+				o.Expect(kerrors.IsForbidden(err)).To(o.BeTrue())
```
Comment (enj) [inline test/extended/authorization/rolebinding_restrictions.go]: Check the string more specifically.

File: test/extended/authorization/rolebinding_restrictions.go [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+				ns := oc.CreateProject()
+				user1 := "dave"
+				user2 := "eve"
+				_, err := oc.AdminAuthorizationClient().Authorization().Roles(ns).Create(generateRole(ns))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindings(ns).Create(generateRolebinding(ns, user1, "rb1"))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindingRestrictions(ns).Create(generateAllowUserRolebindingRestriction(ns, user1))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindings(ns).Create(generateRolebinding(ns, user2, "rb2"))
```
Comment (enj) [inline test/extended/authorization/rolebinding_restrictions.go]: This will flake in CI

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go [hunk 16]
```diff
@@ -239,11 +248,15 @@ func (i *KubeAPIServerInformers) GetOpenshiftSecurityInformers() securityv1infor
 func (i *KubeAPIServerInformers) GetOpenshiftUserInformers() userinformer.SharedInformerFactory {
 	return i.OpenshiftUserInformers
 }
+func (i *KubeAPIServerInformers) GetOpenshiftAuthzInformers() authzinformer.SharedInformerFactory {
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: https://github.com/openshift/origin/pull/22783/files#diff-01a8de689f656fdb0f202987ee6c1f2cR117   ok? They're namespaced so the 
openshift/client-go/authorization/informers/externalversions/authorization/v1/  RoleBindingRestrictionInformer won't work here (I might be missing something, might be unclear on what you're asking here)

File: test/extended/authorization/rolebinding_restrictions.go [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+	rbacv1 "k8s.io/api/rbac/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/apis/authorization"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+//var _ = g.Describe("[Feature: RoleBinding Restrictions][Serial] RoleBindingRestrictions should be functional", func() {
```
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: done

File: test/extended/authorization/rolebinding_restrictions.go [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+var _ = g.Describe("[Feature: RoleBinding Restrictions] RoleBindingRestrictions should be functional", func() {
+	defer g.GinkgoRecover()
+	oc := exutil.NewCLI("rolebinding-restrictions", exutil.KubeConfigPath())
+	g.Context("", func() {
```
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: yes, based on the other extended tests. (cd test; ack g.Context)

File: test/extended/authorization/rolebinding_restrictions.go [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+		g.Describe("Create a rolebinding when there are no restrictions", func() {
+			g.It(fmt.Sprintf("should succeed"), func() {
+				ns := oc.CreateProject()
+				user := "alice"
+				_, err := oc.AdminAuthorizationClient().Authorization().Roles(ns).Create(generateRole(ns))
```
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: i guess not, I've removed that

File: test/extended/authorization/rolebinding_restrictions.go [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+		g.Describe("Create a rolebinding when subject is already bound", func() {
+			g.It(fmt.Sprintf("should succeed"), func() {
+				user := "cindy"
+				ns := oc.CreateProject()
+				_, err := oc.AdminAuthorizationClient().Authorization().Roles(ns).Create(generateRole(ns))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindings(ns).Create(generateRolebinding(ns, user, "rb1"))
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.AdminAuthorizationClient().Authorization().RoleBindingRestrictions(ns).Create(generateAllowUserRolebindingRestriction(ns, user))
```
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: suggestion welcome : )  Haven't seen this fail

File: test/extended/authorization/rolebinding_restrictions.go (hunk 14 above)
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: IsForbidden not enough?

File: test/extended/authorization/rolebinding_restrictions.go (hunk 15 above)
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: it hasn't _yet_? I've run it like 50 times since opening this PR. hm... 
suggestions on how to make it un-flaky?

File: test/extended/authorization/rolebinding_restrictions.go (hunk 14 above)
Comment (enj) [reply test/extended/authorization/rolebinding_restrictions.go]: You can get a forbidden error for a lot of different reasons.  Hence I want both a forbidden check and a string check to confirm the reason is RBR related.

File: test/extended/authorization/rolebinding_restrictions.go (hunk 15 above)
Comment (enj) [reply test/extended/authorization/rolebinding_restrictions.go]: You are racing the watch that is feeding the server side informer meaning you will flake every so often.  `wait.PollImmediate` is probably the best you can do.

File: test/extended/authorization/rolebinding_restrictions.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (enj) [reply test/extended/authorization/rolebinding_restrictions.go]: see below

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (enj) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: We cannot use this authorization informer for anything else because it is JSON based.  Hence it needs to be RBR only.  Something like:

`func (i *KubeAPIServerInformers) GetOpenshiftRoleBindingRestrictionInformer() RoleBindingRestrictionInformer {`

[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: done

File: test/extended/authorization/rolebinding_restrictions.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: ok, i used UntilWithSync

File: test/extended/authorization/rolebinding_restrictions.go (hunk 14 above)
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: added the string check

File: test/extended/authorization/rolebinding_restrictions.go (hunk 15 above)
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: ok used UntilWithSync (I _think_ I used it correctly)

**Comment by sallyom** [comment]
tests passed locally, retesting

File: test/extended/authorization/rolebinding_restrictions.go [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/apis/authorization"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+var _ = g.Describe("[Feature: RoleBinding Restrictions] RoleBindingRestrictions should be functional", func() {
+	defer g.GinkgoRecover()
+	var (
+		oc      = exutil.NewCLI("rolebinding-restrictions", exutil.KubeConfigPath())
+		cleanup = func(ns string) {
```
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: _probably_  but I've since removed this cleanup, since I was cleaning up users created by test framework that I did not create, and after asking around in slack, determined that i should not be cleaning those up (or, the cleanup needs to happen elsewhere)  you'll notice if you run e2es locally after if you `oc get users` you'll see a slew of (non-idp)users/uids leftover.

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go [hunk 22]
```diff
@@ -104,14 +106,15 @@ func NewOpenShiftKubeAPIServerConfigPatch(delegateAPIServer genericapiserver.Del
...
-			UserInformers:                kubeAPIServerInformers.GetOpenshiftUserInformers(),
+			DefaultNodeSelector:            kubeAPIServerConfig.ProjectConfig.DefaultNodeSelector,
+			OriginQuotaRegistry:            quotaRegistry,
+			RESTClientConfig:               *genericConfig.LoopbackClientConfig,
+			ClusterResourceQuotaInformer:   kubeAPIServerInformers.GetOpenshiftQuotaInformers().Quota().V1().ClusterResourceQuotas(),
+			ClusterQuotaMapper:             clusterQuotaMappingController.GetClusterQuotaMapper(),
+			RegistryHostnameRetriever:      registryHostnameRetriever,
+			SecurityInformers:              kubeAPIServerInformers.GetOpenshiftSecurityInformers().Security().V1().SecurityContextConstraints(),
+			UserInformers:                  kubeAPIServerInformers.GetOpenshiftUserInformers(),
+			RoleBindingRestrictionInformer: kubeAPIServerInformers.GetOpenshiftRoleBindingRestrictionInformer().Authorization().V1().RoleBindingRestrictions(),
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: done

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go [hunk 23]
```diff
@@ -240,11 +249,15 @@ func (i *KubeAPIServerInformers) GetOpenshiftSecurityInformers() securityv1infor
 func (i *KubeAPIServerInformers) GetOpenshiftUserInformers() userinformer.SharedInformerFactory {
 	return i.OpenshiftUserInformers
 }
+func (i *KubeAPIServerInformers) GetOpenshiftRoleBindingRestrictionInformer() authzinformer.SharedInformerFactory {
+	return i.OpenshiftRoleBindingRestrictionInformer
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/patch.go]: done

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go [hunk 24]
```diff
@@ -46,6 +46,7 @@ var (
 		"quota.openshift.io/ClusterResourceQuota",
 		"security.openshift.io/SecurityContextConstraint",
 		"security.openshift.io/SCCExecRestrictions",
+		"authorization.openshift.io/RestrictSubjectBindings",
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: done

File: pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go [hunk 25]
```diff
@@ -46,6 +46,7 @@ var (
 		"quota.openshift.io/ClusterResourceQuota",
 		"security.openshift.io/SecurityContextConstraint",
 		"security.openshift.io/SCCExecRestrictions",
+		//"authorization.openshift.io/RestrictSubjectBindings",
 	)
```
Comment (sallyom) [inline pkg/cmd/openshift-kube-apiserver/kubeadmission/register.go]: still stuck on how to fix test/cmd & integration without this

File: test/extended/authorization/rolebinding_restrictions.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (sallyom) [reply test/extended/authorization/rolebinding_restrictions.go]: @enj as you can notice my git history got fubar, so I ended up taking this file from your merged/not-merged PR, will fix

**Comment by sallyom** [comment]
@enj thanks for your help in fixing test/cmd and integration : ) sorry i squashed your commits into mine
